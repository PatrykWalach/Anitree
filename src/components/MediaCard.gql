#import './MediaCardActions.gql'
#import './MediaCardBanner.gql'
#import './MediaCardItem.gql'
#import './MediaCardStatus.gql'

fragment MediaCard_media on Media {
  id
  bannerImage
  ...MediaCardItem_media
  ...MediaCardActions_media
  ...MediaCardBanner_media
  ...MediaCardStatus_media
}
