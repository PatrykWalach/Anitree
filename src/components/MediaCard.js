import { MediaCardActions_media } from './MediaCardActions.js'
import { MediaCardBanner_media } from './MediaCardBanner.js'
import { MediaCardItem_media } from './MediaCardItem.js'
import { MediaCardStatus_media } from './MediaCardStatus.js'

import gql from 'graphql-tag'

export const MediaCard_media = gql`
  fragment MediaCard_media on Media {
    id
    bannerImage
    ...MediaCardItem_media
    ...MediaCardActions_media
    ...MediaCardBanner_media
    ...MediaCardStatus_media
  }
  ${MediaCardItem_media}
  ${MediaCardActions_media}
  ${MediaCardBanner_media}
  ${MediaCardStatus_media}
`
