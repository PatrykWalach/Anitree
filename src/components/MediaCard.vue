<template>
  <ApolloQuery
    v-slot="{ result: { error, data }, isLoading, query }"
    :query="require('@/graphql/queries/Media.gql')"
    :variables="variables"
    :tag="null"
  >
    <v-banner v-if="error" icon="warning" two-line :tile="false">
      There was an error while loading the data. Please make sure your have a
      stable internet connection and try again.

      <template v-slot:actions>
        <v-btn color="error" text @click="query.refetch()">Retry</v-btn>
      </template>
    </v-banner>
    <v-card v-else>
      <MediaCardBanner :media="data && data.Media" />

      <v-tooltip top>
        <template v-slot:activator="hover">
          <!-- <div v-on="on" v-bind="attrs"> -->
          <MediaCardTabs :hover="hover" :media="data && data.Media" />
          <!-- </div> -->
        </template>
        <span>{{ title(data && data.Media.title) }}</span>
      </v-tooltip>

      <v-divider class="mx-4"></v-divider>
      <MediaCardActions :media="data && data.Media" />
      <MediaCardStatus :media="data && data.Media" />
    </v-card>
  </ApolloQuery>
</template>
<script lang="ts">
import MediaCardBanner from './MediaCardBanner.vue'
import MediaCardStatus from './MediaCardStatus.vue'
import MediaCardTabs from './MediaCardTabs.vue'
import MediaCardActions from './MediaCardActions.vue'

import { Variables } from '@/graphql/schema/media'

interface RawBindings {}

import { createComponent, Ref, computed } from '@vue/composition-api'
import useTitle from '@/store/title'

export default createComponent<Readonly<Props>>({
  setup(props) {
    const variables: Ref<Variables> = computed(() => {
      return { id: props.id }
    })
    const { title } = useTitle()

    return { variables, title }
  },
  components: {
    MediaCardBanner,
    MediaCardTabs,
    MediaCardStatus,
    MediaCardActions
  },
  props: {
    id: {
      type: Number,
      required: true,
      default: 2
    }
  }
})
export interface Props {
  id: number
}
</script>

<!--
<script lang="tsx">
import MediaCardBanner from './MediaCardBanner.vue'
import MediaCardStatus from './MediaCardStatus.vue'
import MediaCardTabs from './MediaCardTabs.vue'
import MediaCardActions from './MediaCardActions.vue'
import { VDivider, VBtn, VBanner, VCard } from 'vuetify/lib'

import { MEDIA } from '@/graphql'
import { Media } from '@/graphql/schema/media'

import { createElement } from '@vue/composition-api'

export default createComponent<Readonly<Props>>({
  setup: () => {
    const h = createElement

    return (props ) => {
      const card = (media: Media | null) => (
        <VCard>
          <MediaCardBanner media={media} />
          <MediaCardTabs media={media} />
          <VDivider class="mx-4"></VDivider>
          <MediaCardActions media={media} />
          <MediaCardStatus media={media} />
        </VCard>
      )

      const banner = query => (
        <VBanner
          icon="warning"
          two-line
          tile={false}
          scopedSlots={{
            actions: () => (
              <VBtn color="error" text onClick={query.refetch()}>
                Retry
              </VBtn>
            )
          }}
        >
          There was an error while loading the data. Please make sure your have
          a stable internet connection and try again.
        </VBanner>
      )

      return (
        <ApolloQuery
          scopedSlots={{
            default: ({ result: { error, data }, query }) =>
              error ? banner(query) : card(data && data.Media)
          }}
          query={MEDIA}
          variables={{ id: props.id }}
          tag={null}
        ></ApolloQuery>
      )
    }
  },
  props: {
    id: {
      type: Number,
      required: true
    }
  } 
})

export interface Props {
  id: number
}
</script>
-->
