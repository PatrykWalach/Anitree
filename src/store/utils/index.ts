export * from './useCommands'
export * from './useTitle'
export * from './useEdit'
