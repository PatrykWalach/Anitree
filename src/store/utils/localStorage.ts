export const getItem = <T>(key: string, defaultValue?: T) => {
  const cache = localStorage.getItem(key)
  const stored = cache && JSON.parse(cache)

  return {
    ...defaultValue,
    ...stored,
  } as T
}
export const setItem = <T>(key: string, value: T) =>
  localStorage.setItem(key, JSON.stringify({ ...value }))
