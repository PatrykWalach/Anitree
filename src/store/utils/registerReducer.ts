// import * as reducers from '../reducers'
// import { Reducer, ReducersMapObject, combineReducers } from 'redux'
// import { EnhancedStore } from '@reduxjs/toolkit'
// import { store, useStore } from '../'

// const staticReducers: ReducersMapObject<any, any> = reducers
// const asyncReducers: ReducersMapObject<any, any> = {}

// export const registerReducer = <K extends string, R extends Reducer<any>>(
//   key: K,
//   asyncReducer: R,
// ) => {
//   asyncReducers[key] = asyncReducer
//   store.replaceReducer(createReducer(asyncReducers) as any)

//   return () =>
//     useStore<
//       typeof store extends EnhancedStore<infer S>
//         ? EnhancedStore<S & { [key in K]: R }>
//         : any
//     >()
// }

// export const unregisterReducer = <K extends string>(key: K) => {
//   delete asyncReducers[key]
//   store.replaceReducer(createReducer(asyncReducers) as any)
//   return () =>
//     useStore<
//       typeof store extends EnhancedStore<infer S>
//         ? EnhancedStore<Omit<S, K>>
//         : any
//     >()
// }

// const createReducer = (asyncReducers: ReducersMapObject<any, any> = {}) =>
//   combineReducers({
//     ...staticReducers,
//     ...asyncReducers,
//   })
