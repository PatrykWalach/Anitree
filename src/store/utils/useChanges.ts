import {
  DELETE_MEDIA_LIST_ENTRY,
  MEDIA,
  SAVE_MEDIA_LIST_ENTRY,
} from '@/graphql'
import { DeletePending, SavePending } from '../modules/changes'
import {
  DeleteVariables,
  Deleted,
  SaveVariables,
} from '@/graphql/schema/listEntry'
import { DocumentNode, FetchResult } from 'apollo-link'
import { Ref, ref } from '@vue/composition-api'
import { DataProxy } from 'apollo-cache'
import { Media } from '@/graphql/schema/media'
import { MediaList } from '@/graphql/schema/mediaListCollection'
import { MutationUpdaterFn } from 'apollo-client'
import { Variables } from '@/graphql/schema/media'
import { useActions } from '..'
import { useDispatch } from 'vue-redux-hooks'
import { useMutation } from '@vue/apollo-composable'

type OnError = (
  fn: (param?: Error | undefined) => void,
) => {
  off: () => void
}

export const handleError = <
  P extends Omit<SavePending, 'id'> | Omit<DeletePending, 'id'>
>(
  onError: OnError,
  type: P['type'],
) => {
  const pending: Ref<P> = ref({ type })

  const dispatch = useDispatch()
  const actions = useActions()

  onError(() => {
    dispatch(actions.changes.UNSHIFT_PENDING(pending.value))
  })

  return pending
}

export const useChange = <D, V>(
  query: DocumentNode,
  update: (cache: DataProxy, { data }: FetchResult<D>, mediaId: number) => void,
) => {
  const { mutate: change, loading, onError, onDone } = useMutation<D, V>(query)
  const mutate = (variables: V, mediaId: number) => {
    change(variables, {
      update: (cache, result) => update(cache, result, mediaId),
    })
  }
  return { loading, mutate, onDone, onError }
}

export const useDelete = () => {
  return useChange<{ DeleteMediaListEntry: Deleted }, DeleteVariables>(
    DELETE_MEDIA_LIST_ENTRY,
    updateDeleteMediaListEntry,
  )
}

export const useSave = () => {
  return useChange<
    {
      SaveMediaListEntry: MediaList
    },
    SaveVariables
  >(SAVE_MEDIA_LIST_ENTRY, updateSaveMediaListEntry)
}

export const useDeleteWithChanges = () => {
  const query = useDelete()

  const pending = handleError(query.onError, 'DELETE')

  const mutate = (variables: DeleteVariables, mediaId: number) => {
    pending.value = {
      mediaId,
      type: 'DELETE',
      variables,
    }

    query.mutate(variables, mediaId)
  }

  return { ...query, mutate }
}
export const useSaveWithChanges = () => {
  const query = useSave()

  const pending = handleError(query.onError, 'SAVE')

  const mutate = (variables: SaveVariables) => {
    pending.value = {
      mediaId: variables.mediaId,
      type: 'SAVE',
      variables,
    }

    query.mutate(variables, variables.mediaId)
  }

  return { ...query, mutate }
}

export const useChanges = () => {
  const { mutate: deleteEntry } = useDelete()
  const { mutate: saveEntry } = useSave()

  const isSavePending = (
    pending: SavePending | DeletePending,
  ): pending is SavePending => pending.type === 'SAVE'

  return { deleteEntry, isSavePending, saveEntry }
}

export const updateSaveMediaListEntry: MutationUpdaterFn<{
  SaveMediaListEntry: MediaList
}> = (cache, { data: result }) => {
  if (!result) return

  const query: {
    variables: Variables
    query: DocumentNode
  } = {
    query: MEDIA,
    variables: { id: result.SaveMediaListEntry.mediaId },
  }

  const data: { Media: Media } | null = cache.readQuery(query)

  if (data) {
    data.Media.mediaListEntry = result.SaveMediaListEntry

    cache.writeQuery({
      ...query,
      data,
    })
  }
}

const updateDeleteMediaListEntry = (
  cache: DataProxy,
  { data: result }: FetchResult<{ DeleteMediaListEntry: Deleted }>,
  id: number,
) => {
  if (!result || !result.DeleteMediaListEntry.deleted) return

  const query: {
    variables: Variables
    query: DocumentNode
  } = {
    query: MEDIA,
    variables: {
      id,
    },
  }

  const data: { Media: Media } | null = cache.readQuery(query)

  if (data) {
    data.Media.mediaListEntry = null

    cache.writeQuery({
      ...query,
      data,
    })
  }
}
