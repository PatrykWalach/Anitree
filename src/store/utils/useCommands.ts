import { ListCommand } from '../modules/commands'
import { computed } from '@vue/composition-api'
import { useState } from './useState'
import { useStore } from './useStore'

export const useCommands = () => {
  const { dispatch, actions } = useStore()

  const history = useState(state => state.commands.history)

  const undo = async () => {
    history.value[history.value.length - 1]
      .undo()
      .then(() => dispatch(actions.commands.POP_COMMAND()))
  }

  const add = async (command: ListCommand) => {
    dispatch(actions.commands.PUSH_COMMAND(command))
    await command.execute()
    return command
  }

  const getPending = computed(() => history.value.filter(({ done }) => !done))
  return { add, getPending, undo }
}
