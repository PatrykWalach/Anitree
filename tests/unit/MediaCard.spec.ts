import MediaCard from '@/components/MediaCard.vue'

import { matchSnapshot } from './utils'

describe('MediaCard', () => {
  matchSnapshot(MediaCard, {
    propsData: {
      id: 1,
    },
  })
})
