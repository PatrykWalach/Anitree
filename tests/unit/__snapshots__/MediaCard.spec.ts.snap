// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaCard matches snapshot 1`] = `
<basequery-stub
  apollo="[object Object]"
/>
`;
